### PR TITLE
Allow unschedulable sessions

### PR DESCRIPTION
**Pull request by jcoyne**
Sessions that are entered before the scheduler are run are now
immutable with regards to their timeslot. They will not get scheduled
elsewhere.  This enable you to add sessions that have to be in a
certain place at a certain time. The scheduler pays no-attention to them,
so you only want to place them into Timeslots where schedulable is
false.  For example you may place a "Presenters Luncheon, rm 123"
during the  "Lunch" timeslot.

Also added a rake task `app:clear_schedule` This enables the generated
schedule to be cleared so an alternate schedule can be generated.

### CODE REVIEW

**Comment by pcantrell** [comment]
Good job wading into my tangled scheduler code!

The usual workflow for the scheduling is to generate a schedule on somebody's local machine (which is much, much faster than heroku), review it, then export it from the laptop & import it.

For that reason, `app:clear_schedule` has limited utility. I'd be inclined to keep it a manual operation, given how dangerous it is. If we do include it, I'd recommend including some kind of sanity check (e.g. “type SCHEDULE ARMAGEDDON to confirm”).

**Comment by jcoyne** [comment]
@pcantrell Good advice. I've added a sanity check.

**Comment by pcantrell** [comment]
Ha, awesome. I meant “SCHEDULE ARMAGEDDON” just as a placeholder, but I guess it does catch one’s attention!

**Comment by jcoyne** [comment]
@pcantrell I can recognize a golden idea when I see it.

File: src/lib/scheduling/context.rb [hunk 1]
```diff
@@ -8,11 +8,11 @@ module Scheduling
...
   #
   class Context
-    
+
     attr_reader :sessions, :timeslots, :rooms
-  
+
     def initialize(event)
-      @sessions = event.session_ids
+      @sessions = event.sessions.where(timeslot: nil).select(:id).map(&:id)
```
Comment (caseyhelbling) [inline src/lib/scheduling/context.rb]: Normally I'd just use 
```
  @sessions = event.sessions.where(timeslot: nil).pluck('id')
```

File: src/lib/tasks/app.rake [hunk 2]
```diff
@@ -131,19 +131,38 @@ namespace :app do
...
-  desc 'create a schedule for most recent event'
+
+  desc 'clear the current schedule (DANGER: Irreversible!!!). You must do the before generating a new schedule'
+  task clear_schedule: :environment do
+    STDOUT.puts "Are you sure? This destroys the existing schedule and you will not be\n" \
+      "able to retrieve it. You should back up the database before doing this.\n\nIf you are really sure, type \"SCHEDULE ARMAGEDDON\" now (anything else to cancel)..."
+    input = STDIN.gets.strip
+    if input == 'SCHEDULE ARMAGEDDON'
+      event = Event.current_event
+      event.sessions.each { |s| s.update(timeslot_id: nil) }
```
Comment (caseyhelbling) [inline src/lib/tasks/app.rake]: A little more straight forward might be to use `update_all`

```
Event.current_event.sessions.update_all("timeslot_id = NULL")
```
or whatever the appropriate syntax. The loop is fine but might take a second or two.

File: src/lib/scheduling/context.rb (hunk 1 above)
Comment (jcoyne) [inline src/lib/scheduling/context.rb]: Yeah. That is better.

File: src/lib/tasks/app.rake (hunk 2 above)
Comment (jcoyne) [inline src/lib/tasks/app.rake]: Agreed. that is better.

**Pull request merged**